File: quarry/types/events.go
package types

// ContractVersion is the emit contract version per CONTRACT_EMIT.md.
const ContractVersion = "0.1.0"

// EventType represents the type of event per CONTRACT_EMIT.md.
type EventType string

// Event type constants per CONTRACT_EMIT.md.
const (
	EventTypeItem        EventType = "item"
	EventTypeArtifact    EventType = "artifact"
	EventTypeCheckpoint  EventType = "checkpoint"
	EventTypeEnqueue     EventType = "enqueue"
	EventTypeRotateProxy EventType = "rotate_proxy"
	EventTypeLog         EventType = "log"
	EventTypeRunError    EventType = "run_error"
	EventTypeRunComplete EventType = "run_complete"
)

// IsTerminal returns true if this event type is a terminal event.
func (e EventType) IsTerminal() bool {
	return e == EventTypeRunComplete || e == EventTypeRunError
}

// LogLevel represents log severity per CONTRACT_EMIT.md.
type LogLevel string

// Log level constants per CONTRACT_EMIT.md.
const (
	LogLevelDebug LogLevel = "debug"
	LogLevelInfo  LogLevel = "info"
	LogLevelWarn  LogLevel = "warn"
	LogLevelError LogLevel = "error"
)

// EventEnvelope is the envelope for all events per CONTRACT_EMIT.md.
// All fields use msgpack tags to match the TypeScript SDK wire format.
type EventEnvelope struct {
	// ContractVersion is the semantic version of the emit contract.
	ContractVersion string `msgpack:"contract_version"`
	// EventID is a unique identifier for this event, scoped to the run.
	EventID string `msgpack:"event_id"`
	// RunID is the canonical run identifier.
	RunID string `msgpack:"run_id"`
	// Seq is the monotonic sequence number, starts at 1.
	Seq int64 `msgpack:"seq"`
	// Type is the event type discriminator.
	Type EventType `msgpack:"type"`
	// Ts is the event timestamp in ISO 8601 UTC format.
	Ts string `msgpack:"ts"`
	// Payload is the type-specific payload.
	Payload map[string]any `msgpack:"payload"`
	// JobID is the job identifier, included when known.
	JobID *string `msgpack:"job_id,omitempty"`
	// ParentRunID is the parent run ID for retries.
	ParentRunID *string `msgpack:"parent_run_id,omitempty"`
	// Attempt is the attempt number, always present, starts at 1.
	Attempt int `msgpack:"attempt"`
}

// ItemPayload represents an item event payload per CONTRACT_EMIT.md.
type ItemPayload struct {
	// ItemType is a caller-defined type label.
	ItemType string `msgpack:"item_type"`
	// Data is the record payload.
	Data map[string]any `msgpack:"data"`
}

// ArtifactPayload represents an artifact event payload per CONTRACT_EMIT.md.
// This is the commit record for an artifact; bytes are transmitted separately.
type ArtifactPayload struct {
	// ArtifactID is the unique identifier for the artifact.
	ArtifactID string `msgpack:"artifact_id"`
	// Name is a human-readable name.
	Name string `msgpack:"name"`
	// ContentType is the MIME content type.
	ContentType string `msgpack:"content_type"`
	// SizeBytes is the total size in bytes.
	SizeBytes int64 `msgpack:"size_bytes"`
}

// CheckpointPayload represents a checkpoint event payload per CONTRACT_EMIT.md.
type CheckpointPayload struct {
	// CheckpointID is the unique identifier for the checkpoint.
	CheckpointID string `msgpack:"checkpoint_id"`
	// Note is an optional human-readable note.
	Note *string `msgpack:"note,omitempty"`
}

// EnqueuePayload represents an enqueue event payload per CONTRACT_EMIT.md.
// This is advisory only; not guaranteed or required.
type EnqueuePayload struct {
	// Target is the target identifier for the work.
	Target string `msgpack:"target"`
	// Params is the parameters for the work.
	Params map[string]any `msgpack:"params"`
}

// RotateProxyPayload represents a rotate_proxy event payload per CONTRACT_EMIT.md.
// This is advisory only; not guaranteed or required.
type RotateProxyPayload struct {
	// Reason is an optional reason for the rotation request.
	Reason *string `msgpack:"reason,omitempty"`
}

// LogPayload represents a log event payload per CONTRACT_EMIT.md.
type LogPayload struct {
	// Level is the log level.
	Level LogLevel `msgpack:"level"`
	// Message is the log message.
	Message string `msgpack:"message"`
	// Fields is optional structured fields.
	Fields map[string]any `msgpack:"fields,omitempty"`
}

// RunErrorPayload represents a run_error event payload per CONTRACT_EMIT.md.
type RunErrorPayload struct {
	// ErrorType is the error type/category.
	ErrorType string `msgpack:"error_type"`
	// Message is the error message.
	Message string `msgpack:"message"`
	// Stack is an optional stack trace.
	Stack *string `msgpack:"stack,omitempty"`
}

// RunCompletePayload represents a run_complete event payload per CONTRACT_EMIT.md.
type RunCompletePayload struct {
	// Summary is an optional summary object.
	Summary map[string]any `msgpack:"summary,omitempty"`
}

File: sdk/src/types/events.ts
/**
 * Event envelope and payload types per CONTRACT_EMIT.md
 */

export const CONTRACT_VERSION = '0.11.0' as const
export type ContractVersion = typeof CONTRACT_VERSION

/** Branded type for run identifiers */
export type RunId = string & { readonly __brand: 'RunId' }

/** Branded type for event identifiers */
export type EventId = string & { readonly __brand: 'EventId' }

/** Branded type for job identifiers */
export type JobId = string & { readonly __brand: 'JobId' }

/** Branded type for artifact identifiers */
export type ArtifactId = string & { readonly __brand: 'ArtifactId' }

/** Branded type for checkpoint identifiers */
export type CheckpointId = string & { readonly __brand: 'CheckpointId' }

/**
 * All supported event types from CONTRACT_EMIT.md
 *
 * Note: 'artifact_chunk' is reserved for IPC framing (CONTRACT_IPC.md)
 * and must never be added to this union.
 */
export type EventType =
  | 'item'
  | 'artifact'
  | 'checkpoint'
  | 'enqueue'
  | 'rotate_proxy'
  | 'log'
  | 'run_error'
  | 'run_complete'

/**
 * Log levels for the log event type.
 */
export type LogLevel = 'debug' | 'info' | 'warn' | 'error'

/**
 * Payload for 'item' events.
 * Represents a durable, structured output record.
 */
export type ItemPayload = {
  /** Caller-defined type label */
  item_type: string
  /** The record payload */
  data: Record<string, unknown>
}

/**
 * Payload for 'artifact' events.
 * Represents a binary or large payload.
 * Note: Actual binary data is handled separately via chunking (IPC layer).
 */
export type ArtifactPayload = {
  /** Unique identifier for the artifact */
  artifact_id: ArtifactId
  /** Human-readable name */
  name: string
  /** MIME content type */
  content_type: string
  /** Total size in bytes */
  size_bytes: number
}

/**
 * Payload for 'checkpoint' events.
 * Represents an explicit script checkpoint.
 */
export type CheckpointPayload = {
  /** Unique identifier for the checkpoint */
  checkpoint_id: CheckpointId
  /** Optional human-readable note */
  note?: string
}

/**
 * Payload for 'enqueue' events.
 * Advisory suggestion to enqueue additional work.
 */
export type EnqueuePayload = {
  /** Target identifier for the work */
  target: string
  /** Parameters for the work */
  params: Record<string, unknown>
  /** Optional source partition override for the child run */
  source?: string
  /** Optional category partition override for the child run */
  category?: string
}

/**
 * Payload for 'rotate_proxy' events.
 * Advisory suggestion to rotate proxy/session identity.
 */
export type RotateProxyPayload = {
  /** Optional reason for the rotation request */
  reason?: string
}

/**
 * Payload for 'log' events.
 * Structured log event emitted by script.
 */
export type LogPayload = {
  /** Log level */
  level: LogLevel
  /** Log message */
  message: string
  /** Optional structured fields */
  fields?: Record<string, unknown>
}

/**
 * Payload for 'run_error' events.
 * Represents a script-level error that should terminate the run.
 */
export type RunErrorPayload = {
  /**
   * Error type/category.
   * Expected values include: script_error, timeout, blocked, abort, etc.
   * Not an exhaustive enum — new error types may be introduced.
   */
  error_type: string
  /** Error message */
  message: string
  /** Optional stack trace */
  stack?: string
}

/**
 * Payload for 'run_complete' events.
 * Represents normal completion of the script.
 */
export type RunCompletePayload = {
  /** Optional summary object */
  summary?: Record<string, unknown>
}

/**
 * Maps event types to their payload types.
 * Used for type-safe payload handling.
 */
export type PayloadMap = {
  item: ItemPayload
  artifact: ArtifactPayload
  checkpoint: CheckpointPayload
  enqueue: EnqueuePayload
  rotate_proxy: RotateProxyPayload
  log: LogPayload
  run_error: RunErrorPayload
  run_complete: RunCompletePayload
}

/**
 * Base envelope fields present on all events.
 * From CONTRACT_EMIT.md.
 *
 * All fields are readonly — envelopes are immutable once constructed.
 */
export type EventEnvelopeBase = {
  /** Semantic version string for the emit contract */
  readonly contract_version: ContractVersion
  /** Unique ID for the event, scoped to a run */
  readonly event_id: EventId
  /** Canonical run identifier */
  readonly run_id: RunId
  /** Monotonic sequence number per run, starts at 1 */
  readonly seq: number
  /** Event timestamp in ISO 8601 UTC */
  readonly ts: string
  /** Job ID, included when known at emit-time */
  readonly job_id?: JobId
  /** Parent run ID, included when run is a retry or child run */
  readonly parent_run_id?: RunId
  /** Attempt number. Always present. Starts at 1. */
  readonly attempt: number
}

/**
 * Typed event envelope with specific event type and payload.
 */
export type EventEnvelope<T extends EventType = EventType> = EventEnvelopeBase & {
  /** Event type identifier */
  readonly type: T
  /** Type-specific payload */
  readonly payload: PayloadMap[T]
}

/**
 * Union type of all possible event envelopes.
 */
export type AnyEventEnvelope = {
  [K in EventType]: EventEnvelope<K>
}[EventType]
